### ISSUE TITLE
Unexpected nan in Bessel function i0

### ISSUE THREAD

**Issue by petschge** [kokkos/kokkos#6475]
When using the Kokkos::Experimental::cyl_bessel_i0 function, some input values lead to unexpected nan values.

I am using nvcc Build cuda_11.8.r11.8/compiler.31833905_0 and g++-11 (Debian 11.3.0-12) 11.3.0

Kokkos version is Release 4.1.00, i.e. commit 1a3ea28f6e97b4c9dd2c8ceed53ad58ed5f94dfe.

cmake a was called as cmake -S kokkos -B build_kokkos -DKokkos_ENABLE_SERIAL=ON -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ENABLE_CUDA_LAMBDA=ON -DKokkos_ENABLE_EXAMPLES=ON -DCMAKE_CXX_COMPILER=/usr/bin/g++-11 -DCMAKE_INSTALL_PREFIX=/home/cabana/install_kokkos

The resulting cmake configure output is:
`
 -- Setting default Kokkos CXX standard to 17
 -- Kokkos version: 4.1.00
 -- The project name is: Kokkos
 -- Using internal gtest for testing
 -- Compiler Version: 11.8.89
 -- kokkos_launch_compiler (/home/cabana/kokkos/bin/kokkos_launch_compiler) is enabled...
 -- Using -std=c++17 for C++17 standard as feature
 -- Built-in Execution Spaces:
 --     Device Parallel: Kokkos::Cuda
 --     Host Parallel: Kokkos::OpenMP
 --       Host Serial: SERIAL
 -- 
 -- Architectures:
 --  TURING75
 -- Using internal desul_atomics copy
 -- Kokkos Devices: OPENMP;CUDA;SERIAL, Kokkos Backends: OPENMP;CUDA;SERIAL
 -- Configuring done
 -- Generating done
 -- Build files have been written to: /home/cabana/build_kokkos
`

Code to reproduce the problem is
```
#include <Kokkos_Vector.hpp>

#define ASSERT_NEAR(val1, val2, abs_err)  assert(fabs((val1)-(val2)) < (abs_err))

int main(int argc, char* argv[]) {
	// Initialize the kokkos runtime
	Kokkos::ScopeGuard scope_guard(argc, argv);

	{
		// Test the Kokkos implementation against values from scipy
		auto _test_kokkos_i0 = KOKKOS_LAMBDA(const int& i) {
			Kokkos::complex out = Kokkos::Experimental::cyl_bessel_i0<Kokkos::complex<double>, double, int>(Kokkos::complex(7.998015e-5, 0.));
			printf("out = %e + %e * i\n", out.real(), out.imag());
			ASSERT_NEAR(out.real(), 1.0000000015992057, 1e-7);
			ASSERT_NEAR(out.imag(), 0.0, 1e-7);
		};
		Kokkos::parallel_for("test kokkos implementation of i0", 1, _test_kokkos_i0);

	}

	return 0;
}
```

**Comment by petschge** [comment]
The KokkosCore_config.h is as follows:

#if !defined(KOKKOS_MACROS_HPP) || defined(KOKKOS_CORE_CONFIG_H)
#error \
    "Do not include KokkosCore_config.h directly; include Kokkos_Macros.hpp instead."
#else
#define KOKKOS_CORE_CONFIG_H
#endif

// KOKKOS_VERSION % 100 is the patch level
// KOKKOS_VERSION / 100 % 100 is the minor version
// KOKKOS_VERSION / 10000 is the major version
#define KOKKOS_VERSION 40100
#define KOKKOS_VERSION_MAJOR 4
#define KOKKOS_VERSION_MINOR 1
#define KOKKOS_VERSION_PATCH 0

/* Execution Spaces */
#define KOKKOS_ENABLE_SERIAL
#define KOKKOS_ENABLE_OPENMP
/* #undef KOKKOS_ENABLE_OPENACC */
/* #undef KOKKOS_ENABLE_OPENMPTARGET */
/* #undef KOKKOS_ENABLE_THREADS */
#define KOKKOS_ENABLE_CUDA
/* #undef KOKKOS_ENABLE_HIP */
/* #undef KOKKOS_ENABLE_HPX */
/* #undef KOKKOS_ENABLE_MEMKIND */
/* #undef KOKKOS_ENABLE_LIBRT */
/* #undef KOKKOS_ENABLE_SYCL */
/* #undef KOKKOS_IMPL_SYCL_DEVICE_GLOBAL_SUPPORTED */

/* General Settings */
#define KOKKOS_ENABLE_CXX17
/* #undef KOKKOS_ENABLE_CXX20 */
/* #undef KOKKOS_ENABLE_CXX23 */

/* #undef KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE */
/* #undef KOKKOS_ENABLE_CUDA_UVM */
#define KOKKOS_ENABLE_CUDA_LAMBDA  // deprecated
/* #undef KOKKOS_ENABLE_CUDA_CONSTEXPR */
/* #undef KOKKOS_ENABLE_IMPL_CUDA_MALLOC_ASYNC */
/* #undef KOKKOS_ENABLE_HIP_RELOCATABLE_DEVICE_CODE */
/* #undef KOKKOS_ENABLE_HIP_MULTIPLE_KERNEL_INSTANTIATIONS */
/* #undef KOKKOS_ENABLE_IMPL_HPX_ASYNC_DISPATCH */
/* #undef KOKKOS_ENABLE_DEBUG */
/* #undef KOKKOS_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK */
/* #undef KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK */
/* #undef KOKKOS_ENABLE_TUNING */
/* #undef KOKKOS_ENABLE_DEPRECATED_CODE_3 */
#define KOKKOS_ENABLE_DEPRECATED_CODE_4
#define KOKKOS_ENABLE_DEPRECATION_WARNINGS
/* #undef KOKKOS_ENABLE_LARGE_MEM_TESTS */
#define KOKKOS_ENABLE_COMPLEX_ALIGN
/* #undef KOKKOS_OPT_RANGE_AGGRESSIVE_VECTORIZATION */
/* #undef KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION */
/* #undef KOKKOS_ENABLE_IMPL_MDSPAN */

/* TPL Settings */
/* #undef KOKKOS_ENABLE_HWLOC */
/* #undef KOKKOS_USE_LIBRT */
/* #undef KOKKOS_ENABLE_HBWSPACE */
#define KOKKOS_ENABLE_LIBDL
/* #undef KOKKOS_ENABLE_LIBQUADMATH */
/* #undef KOKKOS_IMPL_CUDA_CLANG_WORKAROUND */
/* #undef KOKKOS_ENABLE_ONEDPL */

/* #undef KOKKOS_ARCH_SSE42 */
/* #undef KOKKOS_ARCH_ARMV80 */
/* #undef KOKKOS_ARCH_ARMV8_THUNDERX */
/* #undef KOKKOS_ARCH_ARMV81 */
/* #undef KOKKOS_ARCH_ARMV8_THUNDERX2 */
/* #undef KOKKOS_ARCH_A64FX */
/* #undef KOKKOS_ARCH_AMD_AVX2 */
/* #undef KOKKOS_ARCH_AVX */
/* #undef KOKKOS_ARCH_AVX2 */
/* #undef KOKKOS_ARCH_AVX512XEON */
/* #undef KOKKOS_ARCH_KNC */
/* #undef KOKKOS_ARCH_AVX512MIC */
/* #undef KOKKOS_ARCH_POWER7 */
/* #undef KOKKOS_ARCH_POWER8 */
/* #undef KOKKOS_ARCH_POWER9 */
/* #undef KOKKOS_ARCH_INTEL_GEN */
/* #undef KOKKOS_ARCH_INTEL_DG1 */
/* #undef KOKKOS_ARCH_INTEL_GEN9 */
/* #undef KOKKOS_ARCH_INTEL_GEN11 */
/* #undef KOKKOS_ARCH_INTEL_GEN12LP */
/* #undef KOKKOS_ARCH_INTEL_XEHP */
/* #undef KOKKOS_ARCH_INTEL_PVC */
/* #undef KOKKOS_ARCH_INTEL_GPU */
/* #undef KOKKOS_ARCH_KEPLER */
/* #undef KOKKOS_ARCH_KEPLER30 */
/* #undef KOKKOS_ARCH_KEPLER32 */
/* #undef KOKKOS_ARCH_KEPLER35 */
/* #undef KOKKOS_ARCH_KEPLER37 */
/* #undef KOKKOS_ARCH_MAXWELL */
/* #undef KOKKOS_ARCH_MAXWELL50 */
/* #undef KOKKOS_ARCH_MAXWELL52 */
/* #undef KOKKOS_ARCH_MAXWELL53 */
/* #undef KOKKOS_ARCH_PASCAL */
/* #undef KOKKOS_ARCH_PASCAL60 */
/* #undef KOKKOS_ARCH_PASCAL61 */
/* #undef KOKKOS_ARCH_VOLTA */
/* #undef KOKKOS_ARCH_VOLTA70 */
/* #undef KOKKOS_ARCH_VOLTA72 */
#define KOKKOS_ARCH_TURING75
/* #undef KOKKOS_ARCH_AMPERE */
/* #undef KOKKOS_ARCH_AMPERE80 */
/* #undef KOKKOS_ARCH_AMPERE86 */
/* #undef KOKKOS_ARCH_ADA89 */
/* #undef KOKKOS_ARCH_HOPPER */
/* #undef KOKKOS_ARCH_HOPPER90 */
/* #undef KOKKOS_ARCH_AMD_ZEN */
/* #undef KOKKOS_ARCH_AMD_ZEN2 */
/* #undef KOKKOS_ARCH_AMD_ZEN3 */
/* #undef KOKKOS_ARCH_VEGA */
/* #undef KOKKOS_ARCH_VEGA906 */
/* #undef KOKKOS_ARCH_VEGA908 */
/* #undef KOKKOS_ARCH_VEGA90A */
/* #undef KOKKOS_ARCH_NAVI */
/* #undef KOKKOS_ARCH_NAVI1030 */
/* #undef KOKKOS_ARCH_NAVI1100 */

/* #undef KOKKOS_IMPL_32BIT */

**Comment by fnrizzi** [comment]
looking at this

**Comment by fnrizzi** [comment]
@petschge looks like this fails also with gcc

https://godbolt.org/z/zM84oPvne

**Comment by masterleinad** [comment]
Unless, we can figure this out quickly, I suggest we switch to the implementation that also `boost` is using, see https://www.boost.org/doc/libs/1_83_0/libs/math/doc/html/math_toolkit/bessel/mbessel.html.

**Comment by fnrizzi** [comment]
yesi was looking at that too - I don't think they support complex though

**Comment by masterleinad** [comment]
@petschge #6484 should fix this issue.

**Comment by masterleinad** [comment]
Fixed in #6484.
